[charts-pro] Update zoom using `requestAnimationFrame`

JCQuintas opened the pull request:
The idea here is to use `requestAnimationFrame` to trigger to zoom state updates, which makes the entire feature run more smoothly.

This causes some small issues in testing though, where we have to wait the timeouts 😆
We could try using fake timers, but I don't really like that approach, as it was giving me more trouble that it was worth.

alexfauquette commented:
I got a react "Maximum update depth exceeded." error when I run the mouse wheel crassy to zoom the synchronization demo. But I can't reproduce it on [master](https://master--material-ui-x.netlify.app/x/react-charts/zoom-and-pan/) so I assume it comes from this PR. It's the only demo on which I can reproduce the issue

https://github.com/user-attachments/assets/382a82f2-6fde-4762-9820-3969616571fa

JCQuintas commented:
> I got a react "Maximum update depth exceeded." error when I run the mouse wheel crassy to zoom the synchronization demo. But I can't reproduce it on [master](https://master--material-ui-x.netlify.app/x/react-charts/zoom-and-pan/) so I assume it comes from this PR. It's the only demo on which I can reproduce the issue

It happens on https://master--material-ui-x.netlify.app/x/react-charts/zoom-and-pan/ for me as well. It is quite hard to fix that, as debouncing or throttling it would make the zooming out of sync with what the user sees/interacts.

It is also a drawback of controlled zoom, since it will re-render the element, which exacerbate this issue. 

> The animation of line marks are not removed while dragging the chart. In this video light theme is the preview, and dark one is master
> 
>  Capture.video.du.2025-03-26.09-55-56.mp4

Nice, will look into it

JCQuintas commented:
@alexfauquette the issue with the marks is due to react-spring. The problematic code will be removed by https://github.com/mui/mui-x/pull/16961 so the changes that would fix it are not really necessary, as to not add a merge conflict into that one. 

Once that lands the issue will fix itself. 🤷

JCQuintas commented:
> > It happens on [master--material-ui-x.netlify.app/x/react-charts/zoom-and-pan](https://master--material-ui-x.netlify.app/x/react-charts/zoom-and-pan/) for me as well. It is quite hard to fix that, as debouncing or throttling it would make the zooming out of sync with what the user sees/interacts.
> 
> Are you sure? I can't reproduce it on master
> 
> After some experimentation, I found that it's resolved if the `useEnancedEffect` of the props synchronization got replaced by a `useEffect`. Or id the `requestAnimationFrame` is replaced by a `setTimeout(, 0)`

Yeah I tested on both local and deployed versions of master. My mouse scrolls extremely fast, so it is possible it created issues regardless and this change only exacerbated it.

That is a good finding, I'll check it when I'm back. Did you try changing it for a `useLayoutEffect` too?

alexfauquette commented:
> Did you try changing it for a useLayoutEffect too?

What do you mean? For me `useEnhancedEffect` is a `useEffect` on server and a `useLayoutEffect` on client.

JCQuintas commented:
> > Did you try changing it for a useLayoutEffect too?
> 
> What do you mean? For me `useEnhancedEffect` is a `useEffect` on server and a `useLayoutEffect` on client.

I always forget how horribly named `useEnhancedEffect` is 😢

bernardobelchior commented on packages/x-charts/src/internals/plugins/featurePlugins/useChartCartesianAxis/zoom.types.ts:
```diff
@@ -2,7 +2,20 @@
+   */
+  axisId: AxisId;
+  /**
+   * The starting percentage of the zoom range. In the range of 0 to 100.
```
Not sure if is makes sense to mention this

```suggestion
   * The starting percentage of the zoom range. In the range of 0 to 100, inclusive on both ends.
```

JCQuintas replied:
Why would it be exclusive?

bernardobelchior replied:
Just wanted to make sure it was clear. I guess it makes sense that it is inclusive, even if we don't specify it, but it might help clear confusion. For example, `Math.random()` is inclusive of 0, but exclusive of 1 ([source](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random)). 

Not a big deal, though. Up to you

bernardobelchior commented on packages/x-charts/src/internals/plugins/featurePlugins/useChartCartesianAxis/zoom.types.ts:
```diff
@@ -2,7 +2,20 @@
-export type ZoomData = { axisId: AxisId; start: number; end: number };
+export type ZoomData = {
+  /**
+   * The axis id of the zoomed axis.
```
Nit: The first axis is redundant IMO. 

```suggestion
   * The id of the zoomed axis.
```

bernardobelchior commented on packages/x-charts/src/internals/domUtils.ts:
```diff
@@ -148,13 +148,18 @@ export const getStringSize = (text: string | number, style: React.CSSProperties
+        clearTimeout(domCleanTimeout);
+      }
+      domCleanTimeout = setTimeout(() => {
+        // Limit node cleaning to once per render cycle
+        measurementSpan.textContent = '';
```
Do we need to clean it? What's the problem of not cleaning it?

JCQuintas replied:
Else the value is there in the tests when we check for a label value for example. Ok tests we want to immediately remove it as to not pollute test render

bernardobelchior replied:
This line is in the `else` clause, shouldn't that run when we aren't in a test environment?

JCQuintas replied:
Ah nvm. I got this wrong. This behaviour didn't change. Not sure what is the reasoning for cleaning it. @alexfauquette

alexfauquette replied:
The comment from this nice guy might help you ;)

https://github.com/mui/mui-x/pull/14042#issuecomment-2285779789

From what I remembered, the DOM was not cleared from one test to the other, so you had an inivisible span in your test because if you search for the last tick label of the axis, it's probably both in the ticklabel, and the measurement span.

bernardobelchior replied:
My question is: why do we need to clean it outside a test environment? 

Changing `textContent` causes the browser to layout, so we might be able to squeeze some more performance out of this method. No need to change it in this PR, though. Just wanted to know if there are any reasons, so I can assess whether it's useful to investigate.

alexfauquette replied:
outside of test environement, I don't see reasons

bernardobelchior commented on packages/x-charts-pro/src/internals/plugins/useChartProZoom/useChartProZoom.ts:
```diff
@@ -91,43 +92,62 @@ export const useChartProZoom: ChartPlugin<UseChartProZoomSignature> = ({
+  // This is debounced. We want to run it only once after the interaction ends.
+  const removeIsInteracting = React.useMemo(
+    () =>
+      debounce(() =>
```
Shouldn't we define the timeout for the debounce?

JCQuintas replied:
I'll define one, it just has a default I used implicitly

bernardobelchior commented on packages/x-charts-pro/src/internals/plugins/useChartProZoom/useChartProZoom.ts:
```diff
@@ -91,43 +92,62 @@ export const useChartProZoom: ChartPlugin<UseChartProZoomSignature> = ({
-        if (prevState.zoom.isControlled) {
-          return prevState;
-        }
+  // This is throttled. We want to run it at most once per frame.
```
> We want to run it at most once per frame.

If we want this, shouldn't we be using `requestAnimationFrame`? `throttle` uses `setTimeout`, doesn't that mean that it can run more frequently than once per frame?

JCQuintas replied:
Leftover from initial idea. The PR description has the reasoning

bernardobelchior commented:
> It seems we can't use requestAnimationFrame as originally planned. But works with setTimeout.
> My current understanding is that the rAF update change needs to happen and finish before the next rAF callback.

I'm intrigued by this. If we can't process the changes in one frame, how does a `setTimeout` fix the issue? 

I created [this CodeSandbox](https://codesandbox.io/p/sandbox/xr9ddk) to test it and it seems that using `throttle` with a wait of 0 can run more frequently than `requestAnimationFrame`:

<img width="546" alt="image" src="https://github.com/user-attachments/assets/ad0b3371-d724-40f2-82f9-caa47e405214" />

 If that happens and zoom works, why can't we use `requestAnimationFrame`?

bernardobelchior commented:
> `raf` is dependent on the screen frame rate.

Yeah. Why is that problematic? I think that's what I'm missing here: I don't really understand why `requestAnimationFrame` doesn't work here, but `setTimeout` does. 

> Log from my `120hz` screen ![Screenshot 2025-04-02 at 12 57 21](https://private-user-images.githubusercontent.com/9891233/429448539-e410730c-0061-481e-8cde-9b2336c3ecb1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM2MDE4NTYsIm5iZiI6MTc0MzYwMTU1NiwicGF0aCI6Ii85ODkxMjMzLzQyOTQ0ODUzOS1lNDEwNzMwYy0wMDYxLTQ4MWUtOGNkZS05YjIzMzZjM2VjYjEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MDJUMTM0NTU2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzAyYzdjYzhhYTJmOTkzODQ2ZDBkZmNhYjA1NDhiNzAwZjc5NjY2YzllMmZlZjRlZmFhNjZhOTFjYWZjMGFjMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.kNlnqi95-gZwCMsAJdx9uWLRNWuM3YSd0Dt52dZGYWo)

Yeah, you see the same issue. The `setTimeout` is called twice between frames: numbers 6 and 8.

JCQuintas commented:
> Yeah, you see the same issue. The `setTimeout` is called twice between frames: numbers 6 and 8.

True, didn't pick that.

I did more testing and it the issue is only fixed on chrome by using `setTimeout` instead of `requestAnimationFrame`. Safari is still failing 😢

JCQuintas commented:
reverted to using raf again. Moved from `useEnhancedEffect` to `useEffect` in the zoom sync. This seems to fix all the issues we have. 

I didn't go too deep into the `why` though.

bernardobelchior commented on packages/x-charts-pro/src/internals/plugins/useChartProZoom/useChartProZoom.ts:
```diff
@@ -91,43 +90,64 @@ export const useChartProZoom: ChartPlugin<UseChartProZoomSignature> = ({
-      store.update((prev) => ({ ...prev, zoom: { ...prev.zoom, isInteracting } }));
-    },
+  // This is debounced. We want to run it only once after the interaction ends.
+  const removeIsInteracting = React.useMemo(
```
Should we call `removeIsInteracting` in line 76 as well?

bernardobelchior commented:
> reverted to using raf again. Moved from `useEnhancedEffect` to `useEffect` in the zoom sync. This seems to fix all the issues we have.
> 
> I didn't go too deep into the `why` though.

I just spent some time trying to debug this issue and couldn't find an answer either. Sometimes can't even reproduce it. 

I'd be interested to see if https://github.com/mui/mui-x/pull/17233 has any impact, though.


----

Also found an interesting edge case (happens in master as well). When scroll stops because we've hit a limit, there is one more update. You can see that after scroll finishes there's another update that happens after it:



https://github.com/user-attachments/assets/0a427e80-4276-42de-a724-406408322114

JCQuintas commented:
> Also found an interesting edge case (happens in master as well). When scroll stops because we've hit a limit, there is one more update. You can see that after scroll finishes there's another update that happens after it:

Are you scrolling with mouse or macbook trackpad? The trackpad (and some mouses) have `"inertia"`, which causes extra scrolls even after you remove your fingers.

JCQuintas commented:
> I'd be interested to see if https://github.com/mui/mui-x/pull/17233 has any impact, though.

It doesn't, I tested it before 😩

bernardobelchior commented:
> > Also found an interesting edge case (happens in master as well). When scroll stops because we've hit a limit, there is one more update. You can see that after scroll finishes there's another update that happens after it:
> 
> Are you scrolling with mouse or macbook trackpad? The trackpad (and some mouses) have `"inertia"`, which causes extra scrolls even after you remove your fingers.

With a trackpad, yeah. Makes sense 👍

JCQuintas commented on packages/x-internals/src/rafThrottle/rafThrottle.ts:
```diff
@@ -0,0 +1,41 @@
+export interface Cancelable {
+  clear(): void;
```
I just copied it from `throttle`, can make it `cancel` 🤣

JCQuintas replied:
It seems a lot of other apis use the same, `debouce` and `throttle` 😢. Will leave like this for consistency

JCQuintas commented on packages/x-charts-pro/src/ScatterChartPro/ScatterChartPro.zoom.test.tsx:
```diff
@@ -0,0 +1,227 @@
+      fireEvent.wheel(svg, { deltaY: -1, clientX: 50, clientY: 50 });
+      // Wait the animation frame
+      await new Promise((r) => setTimeout(r, 20));
+    }
```
I couldn't make the fake timers work with this behavior 😢

bernardobelchior replied:
Can't we reduce from 200 iterations to a smaller number? Why do we need to so many?

Waiting for 200 animation frames means this loop will take more than 3 seconds. We have 6 of these loops, which means we're artificially adding 18 seconds to our tests. 

We can improve this as a follow-up, though

JCQuintas replied:
200 was a number that made the positioning easier. 

The high number stems from our current implementation of the scroll logic. Since there are a lot of scroll events generated, and using `deltaY` value to actually "zoom" doesn't feel natural to the user (user experience pov), we move the zoom by a static amount every wheel event (throttled by rAF).

The sideeffect of that is that we have to create a high number of scroll events to imitate real scrolling
